test/e2e/specs/editor/various/writing-flow.spec.js

```diff
@@ -0,0 +1,1004 @@
+		await page.keyboard.down( 'Shift' );
+		await pageUtils.pressKeyTimes( 'ArrowLeft', 2 );
+		await page.keyboard.press( 'ArrowUp' );
+		await page.keyboard.up( 'Shift' );
+		await pageUtils.pressKeyWithModifier( 'primary', 'b' );
+
+		await expect.poll( editor.getEditedPostContent )
+			.toBe( `<!-- wp:paragraph -->
+<p>first</p>
+<!-- /wp:paragraph -->
+
+<!-- wp:paragraph -->
+<p>second</p>
+<!-- /wp:paragraph -->` );
+	} );
+
+	test( 'should move to the start of the first line on ArrowUp', async ( {
+		page,
+	} ) => {
+		await page.keyboard.press( 'Enter' );
+		await page.keyboard.type( 'a' );
+
+		async function getHeight() {
+			return await page.evaluate(
+				() => document.activeElement.offsetHeight
+			);
+		}
+
+		const height = await getHeight();
+
+		// Keep typing until the height of the element increases. We need two
+		// lines.
+		while ( height === ( await getHeight() ) ) {
+			await page.keyboard.type( 'a' );
+		}
+
+		// Move to the start of the second line.
+		await page.keyboard.press( 'ArrowLeft' );
+		// Move to the start of the first line.
+		await page.keyboard.press( 'ArrowUp' );
+		// Insert a "." for testing.
+		await page.keyboard.type( '.' );
+
+		// Expect the "." to be added at the start of the paragraph
+		await expect(
+			page.locator( 'role=document[name="Paragraph block"i]' )
+		).toHaveText( /^\.a+$/ );
+	} );
+
+	test( 'should vertically move the caret from corner to corner', async ( {
+		page,
+	} ) => {
+		await page.keyboard.press( 'Enter' );
+		await page.keyboard.type( 'a' );
+
+		async function getHeight() {
+			return await page.evaluate(
+				() => document.activeElement.offsetHeight
+			);
+		}
+
+		const height = await getHeight();
+
+		// Keep typing until the height of the element increases. We need two
+		// lines.
+		while ( height === ( await getHeight() ) ) {
+			await page.keyboard.type( 'a' );
+		}
+
+		// Create a new paragraph.
+		await page.keyboard.press( 'Enter' );
+		// Move to the start of the first line.
+		await page.keyboard.press( 'ArrowUp' );
+		// Insert a "." for testing.
+		await page.keyboard.type( '.' );
+
+		// Expect the "." to be added at the start of the paragraph
+		await expect(
+			page.locator( 'role=document[name="Paragraph block"i]' )
+		).toHaveText( /^a+\.a$/ );
+	} );
+} );
+
+class WritingFlowUtils {
+	constructor( { page } ) {
+		this.page = page;
+
+		this.getActiveBlockName = this.getActiveBlockName.bind( this );
+	}
+
+	async getActiveBlockName() {
+		return await this.page.evaluate(
+			() =>
+				window.wp.data.select( 'core/block-editor' ).getSelectedBlock()
+					?.name
+		);
+	}
+
+	async addDemoContent() {
+		await this.page.keyboard.press( 'Enter' );
+		await this.page.keyboard.type( 'First paragraph' );
+		await this.page.keyboard.press( 'Enter' );
+		await this.page.keyboard.type( '/columns' );
+		await this.page.keyboard.press( 'Enter' );
+		await this.page.click( '[aria-label="Two columns; equal split"]' );
```

Comment:
Could this be written into a role selector?

Reply:
So role selectors are preferred methods in similar cases?

Reply:
Yes, role selectors are always preferred unless they're unavailable.